### ISSUE TITLE
LM Bug: no connect to TLS11 and TLS12 _only_ servers (resolved, patch supplied)

### ISSUE THREAD

**Issue by reichhart** [mcabber/loudmouth#10]
When trying to connect to a server which only speaks TLS 1.1 and TLS 1.2 you won't get a connect:

LM-VERBOSE: Setting up SSL...
LM-SSL: SSL_connect(): 0 / SSL_ERROR_SYSCALL

(0 = EOF / actually no SSL failure)

It is caused by limiting the SSL_method to TLSv1.
This is the patch to fix it:

```
[x@viclavic SPECS]$ cat ../SOURCES/loudmouth-20131107.tls11_tls12.patch
--- loudmouth-20131107.sslerror/loudmouth/lm-ssl-openssl.c      2013-11-03 22:51:26.000000000 +0100
+++ loudmouth-20131107/loudmouth/lm-ssl-openssl.c       2014-11-25 10:27:26.000000000 +0100
@@ -298,7 +298,9 @@
         initialized = TRUE;
     }
 
-    ssl->ssl_method = TLSv1_client_method();
+    /* don't use TLSv1_client_method() because we don't get connections to TLS1_1 and TLS1_2 only servers */
+    /*ssl->ssl_method = TLSv1_client_method();*/
+    ssl->ssl_method = SSLv23_client_method();
     if (ssl->ssl_method == NULL) {
         g_log (LM_LOG_DOMAIN, LM_LOG_LEVEL_SSL,
                "TLSv1_client_method() == NULL");

```

With the patch supplied I get a connect:

LM-VERBOSE: Setting up SSL...
LM-SSL: lm-ssl-openssl.c: Cipher: TLSv1/SSLv3/ECDHE-RSA-AES256-SHA/256
LM-SSL: lm-ssl-openssl.c: SSL_get_verify_result() = 0
LM-SSL: lm-ssl-openssl.c: server = 'jabber.tuxad.net', cn = 'jabber.tuxad.net'

Actually SSLv23_client_method() distinguishes between *all* protocols. It's the usual OpenSSL nonsense... ;-)

The server tested does only speak TLS 1.1 and TLS 1.2:

[root@m-1 ~]# SSLCipherSuite=$(openssl ciphers '!EXPORT:!LOW:ALL:+RC4:@STRENGTH'|tr ":" " ") ./ssltest.sh 81.89.239.233 5269 xmpp
Testing protocols:
SSLv2: NO
SSLv3: NO
TLSv1: NO
TLSv1.1: YES
TLSv1.2: YES
Testing ciphers using protocol TLSv1.2
ECDHE-RSA-AES256-GCM-SHA384 TLSv1.2 Kx=ECDH Au=RSA Enc=AESGCM(256) Mac=AEAD TempKey: ECDH, secp384r1, 384 bits
...

(This can also be verified at the IM Observatory.)

kind regards,
reichhart

**Comment by tuxad-loudmouth** [comment]
See pull request #15:
SSL/TLS protocol usage; use SANs of CERTs

**Comment by tyll** [comment]
afaics this can be closed since pull request #15 was merged.
